bin/e2hs-writer/src/reader.rs

```diff
@@ -0,0 +1,381 @@
+    pub receipts: Receipts,
+}
+
+// This struct reads all blocks in an epoch and creates the block data
+// along with the corresponding proofs required to create an E2HS file.
+pub struct EpochReader {
+    starting_block: u64,
+    ending_block: u64,
+    epoch_accumulator: Option<Arc<EpochAccumulator>>,
+    execution_api: ExecutionApi,
+    era_provider: EraProvider,
+}
+
+impl EpochReader {
+    pub async fn new(
+        epoch: u64,
+        epoch_acc_path: PathBuf,
+        el_provider_url: Url,
+    ) -> anyhow::Result<Self> {
+        let starting_block = epoch * EPOCH_SIZE;
+        let ending_block = starting_block + EPOCH_SIZE;
+        let execution_api = ExecutionApi::new(el_provider_url.clone(), el_provider_url, 10).await?;
+        let latest_block = execution_api.get_latest_block_number().await?;
+        let maximum_epoch = latest_block / EPOCH_SIZE;
+        if epoch > maximum_epoch {
+            bail!("Epoch {epoch} is greater than the maximum epoch {maximum_epoch}");
+        }
+        let era_provider = EraProvider::new(epoch).await?;
+        let epoch_accumulator = match starting_block < MERGE_BLOCK_NUMBER {
+            true => {
+                let epoch_index = starting_block / EPOCH_SIZE;
+                let header_validator = HeaderValidator::new();
+                Some(Arc::new(
+                    lookup_epoch_acc(
+                        epoch_index,
+                        &header_validator.pre_merge_acc,
+                        &epoch_acc_path,
+                    )
+                    .await?,
+                ))
+            }
+            false => None,
+        };
+        Ok(Self {
+            starting_block,
+            ending_block,
+            epoch_accumulator,
+            execution_api,
+            era_provider,
+        })
+    }
+
+    fn get_pre_merge_block_data(&self, block_number: u64) -> anyhow::Result<AllBlockData> {
+        let raw_era1 = self.era_provider.get_era1_for_block(block_number)?;
+        let block_index = block_number % EPOCH_SIZE;
+        let tuple = Era1::get_tuple_by_index(&raw_era1, block_index);
+        let header = tuple.header.header;
+        let receipts = tuple.receipts.receipts;
+        let body = tuple.body.body;
+        let epoch_acc = self.epoch_accumulator.clone().unwrap_or_else(|| {
+            panic!("Epoch accumulator not found for pre-merge block: {block_number}")
+        });
+        let proof = PreMergeAccumulator::construct_proof(&header, &epoch_acc)?;
+        let proof = BlockProofHistoricalHashesAccumulator::new(proof.into()).map_err(|e| {
+            anyhow!("Unable to convert proof to BlockProofHistoricalHashesAccumulator: {e:?}")
+        })?;
+        let header_with_proof = HeaderWithProof {
+            header,
+            proof: BlockHeaderProof::HistoricalHashes(proof),
+        };
+        Ok(AllBlockData {
+            block: block_number,
+            header_with_proof,
+            body,
+            receipts,
```

Comment:
```suggestion
        let epoch_acc = self.epoch_accumulator.clone().unwrap_or_else(|| {
            panic!("Epoch accumulator not found for pre-merge block: {block_number}")
        });
        let proof = PreMergeAccumulator::construct_proof(&header, &epoch_acc)?;
        let proof = BlockProofHistoricalHashesAccumulator::new(proof.into()).map_err(|e| {
            anyhow!("Unable to convert proof to BlockProofHistoricalHashesAccumulator: {e:?}")
        })?;
        let header_with_proof = HeaderWithProof {
            header,
            proof: BlockHeaderProof::HistoricalHashes(proof),
        };
        Ok(AllBlockData {
            block: block_number,
            header_with_proof,
            body: tuple.body.body,
            receipts: tuple.receipts.receipts,
```

These variables aren't used, so I think we shouldn't define them to simplify the code